### ISSUE TITLE
index.php causes silverstripe to fall back to no mod_rewrite 

### ISSUE THREAD

**Issue by torleif** [silverstripe/silverstripe-installer#58]
If you're using mod_rewrite with dynamic URL's index.php causes issues and tries to hijack the urls. To reproduce: 

1) install fresh silverstripe instance with default theme

2) click some links

3) eventually links will become index.php/about, etc. Links stop working and resources (css, js etc) stop loading

Perhaps it might be best to check during install if you're using mod_rewrite and remove/rename index.php.

**Comment by ss23** [comment]
@dhensby This is the installer repo, and that link is a 404 now. Do you recall what it was?

**Comment by dhensby** [comment]
@ss23 I think it was this: https://github.com/h5bp/html5-boilerplate/blob/v5.0.0/dist/.htaccess#L553

**Comment by micschk** [comment]
This seems to be due to some changes to mod_dir in Apache 2.4 (precedence of handlers).

index.php gets added to the URLs as soon as you navigate to the homepage (/), which gets caught as a directory index (and is thus handled by index.php) instead of passed on to mod_rewrite (handled by framework/main.php). After this happens once, all further requests are handled by index.php (which is a file, and therefore doesn't get handled by mod_rewrite anymore).

Further reference from the [Apache docs](http://httpd.apache.org/docs/current/mod/mod_dir.html#DirectoryCheckHandler): 
> In releases prior to 2.4, this module did not take any action if any other handler was configured for a URL. This allows directory indexes to be served even when a SetHandler directive is specified for an entire directory, but it can also result in some conflicts with modules such as mod_rewrite.

It will be fixed by disabling mod_dir in the server config or from the .htaccess (before the mod_rewrite section):
```
<IfModule mod_dir.c>
	DirectoryIndex disabled
</IfModule>
```

**Comment by dhensby** [comment]
@micschk great find. I'll test it locally

**Comment by dhensby** [comment]
Ok, I can confirm this is fixed on Apache 2.4.6 when using

```apache
<IfModule mod_dir.c>
    DirectoryIndex disabled
</IfModule>
```

**Comment by jhnpldng** [comment]
It would be great if someone could put the workaround/solution somewhere on the SS site where it would be easy to find. Like common installation problems.....etc. I beat my head against the wall for a day and a half and almost tossed SS to the curb due to this issue. I found lots of half solutions and guesses in the forum so this IS a common problem. (just ask godaddy) This is one place it could go and I would post it myself but I'm not really up on apache/htaccess so I would probably post poor info.  http://doc.silverstripe.org/en/getting_started/installation/common_problems?_ga=1.51536097.974913142.1423667649

**Comment by micschk** [comment]
@unclecheese, are you by any chance the author of that page? Would it be easy to include the above description & fix?

**Comment by SpiritLevel** [comment]
@dhensby I just did a fresh install of SS (the stable version, 3.1) after a fresh install of Ubuntu 14.04 with a fresh install of lamp-server (apache 2.4.7) and I am having the same problem as the original poster. I have been trying to find the smallest installation that enables SS to run. All errors reported by the SS webinstaller were dealt with (in particular, mod_rewrite was enabled, same with php5-tidy, php5-curl, date.timezone etc.). I checked that the .htaccess has your fix above for the mod_dir module directive and it does. Strangely, on my other machine, with what I think is exactly the same software and configuration, I also tried a fresh install of SS and I don't have this problem! It almost seems that the problem isn't necessarily to do with how mod_dir and mod_rewrite interact but perhaps due to other modules ? Maybe I have an extra apache module running on one of my machines but not the other which the SS web installer doesn't check for ? This would possibly account for the problem occurring in a slightly erratic way (judging by the other posts).

**Comment by SpiritLevel** [comment]
@dhensby There are three extra entries in the mods-enabled directory on my machine that works properly: socache_shmcb.load, ssl.conf, and ssl.load

I'll try adding these to my minimal setup to see if the problem disappears.

**Comment by dhensby** [comment]
@spiritlevel have you tried disabling mod-dir like so: `sudo a2dismod dir`? 

I'm pretty sure this is fixed, so you still having the problem is a little strange. 

If all else fails you could just delete index.php?

**Comment by SpiritLevel** [comment]
@dhensby Disabling mod_dir causes a problem when restarting Apache: conf-enabled/phpmyadmin.conf uses the DirectoryIndex directive and Apache complains. So, I am not too keen to disable mod_dir since this would require changing phpmyadmin.conf. 

Deleting index.php causes the site not to load; the directory is listed instead. Also not an ideal solution!

Anything else I could check ? If helpful, I could give you the exact steps I followed installing everything.

**Comment by dhensby** [comment]
@SpiritLevel have you enabled mod_rewrite?

**Comment by SpiritLevel** [comment]
@dhensby Yup. Just quadruple checked. Can't get past the SS webinstaller with it off.

**Comment by dhensby** [comment]
Odd. OK. If you could create a with gist your steps to provision a server that has the problem, I'll try to take a look

**Comment by SpiritLevel** [comment]
OK :) I'll go through the entire process again, just to be sure, and put it in a gist. I'll let you know when ready to look at. Thanks Dan.

**Comment by SpiritLevel** [comment]
@dhensby OK, steps I took are here:

https://gist.github.com/SpiritLevel/5f9616a65ee11b53f445

**Comment by dhensby** [comment]
@SpiritLevel you haven't enabled mod_rewrite - I was able to replicate the problem and resolved it by running:

```sh
$ sudo a2enmod rewrite
$ sudo service apache2 restart
```

![](https://dl.dropboxusercontent.com/u/3429338/Clippings/2015-08/2015-08-12_11-58-44-mYOYn1Rzqd.png)

**Comment by SpiritLevel** [comment]
@dhensby Thanks for looking at this :) I did enable mod_rewrite ( step (15) ) and restarted apache2 (step (24) ). For good measure, I just now disabled mod_rewrite, restarted the server, then re-enabled mod_rewrite, and restarted the server...but the problem remains. Also, you cannot get through the SS web installer without mod_rewrite being enabled. So, it seems the problem is elsewhere! Did you replicate this with a completely fresh install of everything ? I have been able to reproduce this problem half a dozen times, consistently. My test machine is a slightly older Dell....could there be a hardware problem interfering with apache ?
